File: mvqc/java/Pos.java
import java.util.ArrayList;

public class Pos
{
	public final static int M_UL = 1;
	public final static int M_UR = 2;
	public final static int M_DL = 3;
	public final static int M_DR = 4;
	public final static int[] all_moves = { M_UL, M_UR, M_DL, M_DR };
	public final static int[] down_moves = { M_DL, M_DR };
	public final static int[] up_moves = { M_UL, M_UR };

	public int l, n;

	public Pos(int l, int n) { this.l = l; this.n = n; }
	public String s() { return String.format("%d.%d", l, n); }

	public Pos move(int direction, int board_size)
	{
		int nl;

		if (direction == Pos.M_UR || direction == Pos.M_UL) {
			if (l == 0)
				return null;
			nl = l - 1;
		} else {
			if (l == board_size - 1)
				return null;
			nl = l + 1;
		}

		if (l % 2 == 0) {
			/*
			 * Even line, left move is nr == nr,
			 * right move is +1 and possible for all but last cell
			 */
			if (direction == Pos.M_UL || direction == Pos.M_DL)
				return new Pos(nl, n);
			else if (n < board_size / 2 - 1)
				return new Pos(nl, n + 1);
			else
				return null;
		} else {
			/*
			 * Odd line, right move is nr == nr,
			 * lefot move is -1 and possible for all but 0th cell
			 */
			if (direction == Pos.M_UR || direction == Pos.M_DR)
				return new Pos(nl, n);
			else if (n > 0)
				return new Pos(nl, n - 1);
			else
				return null;
		}
	}

	public ArrayList<Pos> fly_move(int direction, int board_size)
	{
		/* Generate as many moves in one direction as possible */
		ArrayList<Pos> posl = new ArrayList<Pos>();

		Pos pos = this;
		while (true) {
			Pos npos = pos.move(direction, board_size);
			if (npos == null)
				break;

			posl.add(npos);
			pos = npos;
		}

		return posl;
	}
}

File: mvqc/objc/pos.m
#import <Foundation/Foundation.h>
#import "pos.h"

@implementation Pos
-(id)init: (int)line nr: (int) nr
{
	self = [super init];
	l = line;
	n = nr;
	return self;
}

-(Pos *)move: (int)direction board_size: (int)bsiz
{
	int nl;

	if (direction == M_UR || direction == M_UL) {
		if (l == 0)
			return nil;
		nl = l - 1;
	} else {
		if (l == bsiz - 1)
			return nil;
		nl = l + 1;
	}

	if (l % 2 == 0) {
		/*
		 * Even line, left move is nr == nr,
		 * right move is +1 and possible for all but last cell
		 */
		if (direction == M_UL || direction == M_DL)
			return [[Pos alloc] init: nl nr: n];
		else if (n < bsiz / 2 - 1)
			return [[Pos alloc] init: nl nr: n + 1];
		else
			return nil;
	} else {
		/*
		 * Odd line, right move is nr == nr,
		 * lefot move is -1 and possible for all but 0th cell
		 */
		if (direction == M_UR || direction == M_DR)
			return [[Pos alloc] init: nl nr: n];
		else if (n > 0)
			return [[Pos alloc] init: nl nr: n - 1];
		else
			return nil;
	}
}

-(NSMutableArray *)fly_move: (int)direction board_size: (int)bsiz
{
	NSMutableArray *posl = [[NSMutableArray alloc] init];

	Pos *pos = self;
	while (true) {
		Pos *npos = [pos move: direction board_size: bsiz];
		if (npos == nil)
			break;

		[posl addObject: npos];
		[npos release];
		pos = npos;
	}

	return posl;
}
@end

#ifdef DBG
int main( int argc, const char * argv[])
{
	int val;
	Pos *r;
	NSMutableArray *fr;

	NSLog(@"Pos tests");

	/* Regular moves */
	Pos *p = [[Pos alloc] init: 0 nr: 0];
	r = [p move: M_UR board_size: 8];
	if (r != nil) {
		NSLog(@"0.0 UR move");
		return 1;
	}

	r = [p move: M_UL board_size: 8];
	if (r != nil) {
		NSLog(@"0.0 UL move");
		return 1;
	}

	r = [p move: M_DL board_size: 8];
	if (r == nil || r->l != 1 || r->n != 0) {
		NSLog(@"0.0 DL move");
		return 1;
	}

	r = [p move: M_DR board_size: 8];
	if (r == nil || r->l != 1 || r->n != 1) {
		NSLog(@"0.0 DR move");
		return 1;
	}

	p = [[Pos alloc] init: 2 nr: 2];
	r = [p move: M_UR board_size: 8];
	if (r == nil || r->l != 1 || r->n != 3) {
		NSLog(@"2.2 UR move");
		return 1;
	}

	r = [p move: M_UL board_size: 8];
	if (r == nil || r->l != 1 || r->n != 2) {
		NSLog(@"2.2 UL move");
		return 1;
	}

	r = [p move: M_DL board_size: 8];
	if (r == nil || r->l != 3 || r->n != 2) {
		NSLog(@"2.2 DL move");
		return 1;
	}

	r = [p move: M_DR board_size: 8];
	if (r == nil || r->l != 3 || r->n != 3) {
		NSLog(@"0.0 DR move");
		return 1;
	}

	/* Fly moves */
	p = [[Pos alloc] init: 3 nr: 0];

	fr = [p fly_move: M_UL board_size: 8];
	if (fr.count != 0) {
		NSLog(@"3.0 UL fly move");
		return 1;
	}

	fr = [p fly_move: M_UR board_size: 8];
	if (fr.count != 3) {
		NSLog(@"3.0 UR fly move");
		return 1;
	}

	r = [fr objectAtIndex: 0];
	if (r == nil || r->l != 2 || r->n != 0) {
		NSLog(@"3.0 UR fly 0");
		return 1;
	}
	r = [fr objectAtIndex: 1];
	if (r == nil || r->l != 1 || r->n != 1) {
		NSLog(@"3.0 UR fly 1");
		return 1;
	}
	r = [fr objectAtIndex: 2];
	if (r == nil || r->l != 0 || r->n != 1) {
		NSLog(@"3.0 UR fly 2");
		return 1;
	}

	return 0;
}
#endif
